src/irmin-pack/pack_store.ml

```diff
@@ -147,9 +167,108 @@ module Maker (Index : Pack_index.S) (K : Irmin.Hash.S with type t = Index.key) :
       assert (n = K.hash_size);
       decode_bin_hash (Bytes.unsafe_to_string buf) (ref 0)
 
+    type span = { offset : int63; length : int }
+    (** The type of contiguous ranges of bytes in the pack file. *)
+
+    (** Refer to the index for the position of a pack entry, assuming it is
+        indexed: *)
+    let get_entry_span_from_index_exn t hash : span =
+      match index_direct t hash with
+      | Some key' -> (
+          match Pack_key.inspect key' with
+          | Direct { offset; length; _ } -> { offset; length }
+          | Indexed _ ->
+              (* [index_direct] returns only [Direct] keys. *)
+              assert false)
+      | None ->
+          corrupted_store "Unexpected object %a missing from index" pp_hash hash
+
+    module Entry_prefix = struct
+      type t = {
+        hash : hash;
+        kind : Pack_value.Kind.t;
+        value_length : int option;
+            (** Length of the value segment including the length header (if it
+                exists). *)
+      }
+      [@@deriving irmin ~pp_dump]
+
+      let min_length = K.hash_size + 1
+      let max_length = K.hash_size + 1 + Varint.max_encoded_size
+    end
+
+    let io_read_and_decode_entry_prefix ~off t =
+      let buf = Bytes.create Entry_prefix.max_length in
+      let bytes_read = IO.read t.pack.block ~off buf in
+      if bytes_read < Entry_prefix.min_length then
```

Comment:
is this related to https://github.com/mirage/irmin/pull/1534#discussion_r756878195 ? can you maybe add a comment (in the definition of min_length for instance?)

Reply:
Yep, that's the one: the final data section in the pack file may be shorter than 9 bytes (and in the general case it may have length 0). I've added a comment that hopefully clarifies it.

Reply:
thanks!